File: Tests/parallel_loop_reduction_add_loop.f90
      INTEGER FUNCTION test()
        IMPLICIT NONE
        INCLUDE "acc_testsuite.fh"
        INTEGER :: x, y !Iterators
        REAL(8),DIMENSION(10*LOOPCOUNT):: a, b, c, d !Data
        REAL(8) :: avg
        INTEGER :: errors = 0

        !Initilization
        CALL RANDOM_SEED()
        CALL RANDOM_NUMBER(a)
        CALL RANDOM_NUMBER(b)
        avg = 0
        !$acc data copyin(a(1:10*LOOPCOUNT), b(1:10*LOOPCOUNT)) &
            !$acc create(c(1:10*LOOPCOUNT)) &
            !$acc copyout(d(1:10*LOOPCOUNT))
          !$acc parallel loop gang private(avg)
          DO x = 0, 9
            avg = 0
            !$acc loop worker reduction(+:avg)
            DO y = 1, LOOPCOUNT
              avg = avg + a(x * LOOPCOUNT + y) + b(x * LOOPCOUNT + y)
               c(x * LOOPCOUNT + y) = a(x * LOOPCOUNT + y) + b(x * &
                   LOOPCOUNT + y)
            END DO
            avg = avg / LOOPCOUNT
            !$acc loop worker
            DO y = 1, LOOPCOUNT
              d(x * LOOPCOUNT + y) = c(x * LOOPCOUNT + y) / avg
            END DO
          END DO
        !$acc end data
        DO x = 0, 9
          avg = 0
          DO y = 1, LOOPCOUNT
            avg = avg + a(x * LOOPCOUNT + y) + b(x * LOOPCOUNT + y)
          END DO
          avg = avg / LOOPCOUNT
          DO y = 1, LOOPCOUNT
             IF (abs(d(x * LOOPCOUNT + y) - ((a(x * LOOPCOUNT + y) + &
                 b(x * LOOPCOUNT + y)) / avg)) .gt. PRECISION) THEN
              errors = errors + 1
            END IF
          END DO
        END DO
        test = errors
      END


      PROGRAM test_kernels_async_main
      IMPLICIT NONE
      INTEGER :: failed, success !Number of failed/succeeded tests
      INTEGER :: num_tests,crosschecked, crossfailed, j
      INTEGER :: temp,temp1
      INCLUDE "acc_testsuite.fh"
      INTEGER test


      CHARACTER*50:: logfilename !Pointer to logfile
      INTEGER :: result

      num_tests = 0
      crosschecked = 0
      crossfailed = 0
      result = 1
      failed = 0

      !Open a new logfile or overwrite the existing one.
      logfilename = "OpenACC_testsuite.log"
!      WRITE (*,*) "Enter logFilename:"
!      READ  (*,*) logfilename

      OPEN (1, FILE = logfilename)

      WRITE (*,*) "######## OpenACC Validation Suite V 2.5 ######"
      WRITE (*,*) "## Repetitions:", N
      WRITE (*,*) "## Loop Count :", LOOPCOUNT
      WRITE (*,*) "##############################################"
      WRITE (*,*)

      WRITE (*,*) "--------------------------------------------------"
      WRITE (*,*) "Test of parallel_loop_reduction_add_loop"
      WRITE (*,*) "--------------------------------------------------"

      crossfailed=0
      result=1
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*) "Test of parallel_loop_reduction_add_loop"
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*)
      WRITE (1,*) "(Crosstests should fail)"
      WRITE (1,*)

      DO j = 1, N
        temp =  test()
        IF (temp .EQ. 0) THEN
          WRITE (1,*)  j, ". test successfull."
          success = success + 1
        ELSE
          WRITE (1,*) "Error: ",j, ". test failed."
          failed = failed + 1
        ENDIF
      END DO


      IF (failed .EQ. 0) THEN
        WRITE (1,*) "Directive worked without errors."
        WRITE (*,*) "Directive worked without errors."
        result = 0
        WRITE (*,*) "Result:",result
      ELSE
        WRITE (1,*) "Directive failed the test ", failed, " times."
        WRITE (*,*) "Directive failed the test ", failed, " times."
        result = failed * 100 / N
        WRITE (*,*) "Result:",result
      ENDIF
      CALL EXIT (result)
      END PROGRAM

File: Tests/parallel_loop_reduction_add_loop.cpp
#include "acc_testsuite.h"
#ifndef T1
//T1:parallel,loop,reduction,combined-constructs,V:1.0-2.7
int test1(){
    int err = 0;
    srand(SEED);
    real_t * a = new real_t[10 * n];
    real_t* b = new real_t[10 * n];
    real_t* c = new real_t[10 * n];
    real_t* d = new real_t[10 * n];
    real_t avg = 0.0;
    real_t rolling_total = 0;

    for (int x = 0; x < 10 * n; ++x) {
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = rand() / (real_t)(RAND_MAX / 10);
    }

    #pragma acc data copyin(a[0:10*n], b[0:10*n]) create(c[0:10*n]) copyout(d[0:10*n])
    {
        #pragma acc parallel loop gang private(avg)
        for (int x = 0; x < 10; ++x) {
            avg = 0;
            #pragma acc loop worker reduction(+:avg)
            for (int y = 0; y < n; ++y) {
                avg += c[x * n + y] = a[x * n + y] + b[x * n + y];
            }
            avg = avg / n;
            #pragma acc loop worker
            for (int y = 0; y < n; ++y) {
                d[x * n + y] = c[x * n + y] - avg;
            }
        }
    }


    for (int x = 0; x < 10; ++x) {
        rolling_total = 0;
        for (int y = 0; y < n; ++y) {
            rolling_total += a[x * n + y] + b[x * n + y];
        }
        rolling_total = rolling_total / n;
        for (int y = 0; y < n; ++y) {
            if (fabs(d[x * n + y] - ((a[x * n + y] + b[x * n + y]) - rolling_total)) > PRECISION) {
                err += 1;
            }
        }
    }

    return err;
}
#endif

#ifndef T2
//T2:parallel,loop,reduction,combined-constructs,V:2.7-2.7
int test2() {
    int err = 0;
    srand(SEED);
    real_t* a = new real_t[25 * n];
    real_t* b = new real_t[25 * n];
    real_t* c = new real_t[25 * n];
    real_t* d = new real_t[25 * n];
    real_t avgs[5];
    real_t host_avgs[5];

    for (int x = 0; x < 25 * n; ++x) {
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = rand() / (real_t)(RAND_MAX / 10);
        c[x] = 0;
        d[x] = 0;
    }

    #pragma acc data copyin(a[0:25*n], b[0:25*n]) copyout(c[0:25*n], d[0:25*n])
    {
        #pragma acc parallel loop gang private(avgs)
        for (int x = 0; x < 5; ++x) {
            for (int y = 0; y < 5; ++y) {
                avgs[y] = 0;
            }
            #pragma acc loop worker reduction(+:avgs)
            for (int y = 0; y < 5 * n; ++y) {
                avgs[y % 5] += c[x * 5 * n + y] = a[x * 5 * n + y] + b[x * 5 * n + y];
            }
            for (int y = 0; y < 5; ++y) {
                avgs[y] = avgs[y] / n;
            }
            #pragma acc loop worker
            for (int y = 0; y < 5 * n; ++y) {
                d[x * 5 * n + y] = c[x * 5 * n + y] - avgs[y % 5];
            }
        }
    }

    for (int x = 0; x < 5; ++x) {
        for (int y = 0; y < 5; ++y) {
            host_avgs[y] = 0;
        }
        for (int y = 0; y < 5 * n; ++y) {
            host_avgs[y % 5] += a[x * 5 * n + y] + b[x * 5 * n + y];
            if (fabs(c[x * 5 * n + y] - (a[x * n * 5 + y] + b[x * n * 5 + y])) > PRECISION) {
                err += 1;
            }
        }
        for (int y = 0; y < 5; ++y) {
            host_avgs[y] = host_avgs[y] / n;
        }
        for (int y = 0; y < 5 * n; ++y) {
            if (fabs(d[x * 5 * n + y] - (c[x * 5 * n + y] - host_avgs[y % 5])) > PRECISION){
                err += 1;
            
            }
        }
    }

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
#ifndef T2
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test2();
    }
    if (failed != 0){
        failcode = failcode + (1 << 1);
    }
#endif
    return failcode;
}
